Allow funcref in g:LanguageClient_selectionUI

runiq opened the pull request:
Hot on the heels of https://github.com/autozimu/LanguageClient-neovim/pull/1045#issuecomment-641455349, here's a way to use a selection UI other than FZF. The `FZF` selection UI is implemented as a special backwards-compatible version of the `Funcref` selection UI.

The function has to adhere to the arguments provided to `s:FZF()`, i.e. it has to take `source` and a `sink` arguments. [As in `fzf#run()`](https://github.com/junegunn/fzf/blob/master/README-VIM.md#fzfrun), `source` can be either a list or a string, and `sink` may be a string or a reference to another function of one argument (the selected line).

In action, using [vim-clap](https://github.com/liuchengxu/vim-clap) instead of FZF:

![works](https://user-images.githubusercontent.com/620981/85034923-e4295400-b182-11ea-9ac1-5a9852ff0a44.gif)

Config:

```vim
function! MySelectionUI(source, sink) abort
	return clap#run({'id': 'LCN', 'source': a:source, 'sink': a:sink})
endfunction

let g:LanguageClient_selectionUI = function('MySelectionUI')
```

runiq committed:
Allow funcref in g:LanguageClient_selectionUI

Also implement the "FZF" selection UI as a backwards-compatible instance
of a funcref selection UI.

runiq commented on src/language_server_protocol.rs:
```diff
@@ -1652,9 +1652,10 @@ impl LanguageClient {
-            .notify("s:FZF", json!([source, NOTIFICATION__FZFSinkCommand]))?;
+        self.vim()?.rpcclient.notify(
+            "s:selectionUI_funcref",
+            json!([source, NOTIFICATION__FZFSinkCommand]),
```
`rustfmt` changed indentation here; all I did was to replace `s:FZF` with `s:selectionUI_funcref`.

runiq commented on src/language_server_protocol.rs:
```diff
@@ -2060,9 +2061,10 @@ impl LanguageClient {
-            .notify("s:FZF", json!([source?, NOTIFICATION__FZFSinkCommand]))?;
+        self.vim()?.rpcclient.notify(
+            "s:selectionUI_funcref",
+            json!([source?, NOTIFICATION__FZFSinkCommand]),
```
Same as https://github.com/autozimu/LanguageClient-neovim/pull/1059/files#r442321418.

runiq commented:
[In action:](https://github.com/liuchengxu/vim-clap/issues/433#issuecomment-646041694)

![works](https://user-images.githubusercontent.com/620981/85034923-e4295400-b182-11ea-9ac1-5a9852ff0a44.gif)

Config:

```vim
function! MySelectionUI(source, sink) abort
	return clap#run({'id': 'LCN', 'source': a:source, 'sink': a:sink})
endfunction

let g:LanguageClient_selectionUI = function('MySelectionUI')
```

martskins commented on src/language_server_protocol.rs:
```diff
@@ -171,7 +171,7 @@ impl LanguageClient {
         let selectionUI = if let Some(s) = selectionUI {
             SelectionUI::from_str(&s)?
         } else if self.vim()?.eval::<_, i64>("get(g:, 'loaded_fzf')")? == 1 {
```
Should this check for `loaded_fzf` be there? If you are using clap or any other selection UI it doesn't really matter if you have fzf is loaded or not.

martskins commented on autoload/LanguageClient.vim:
```diff
@@ -201,6 +209,14 @@ function! s:inputlist(...) abort
 endfunction
 
+function! s:selectionUI_funcref(source, sink) abort
+    if g:LanguageClient_selectionUI ==? 'FZF'
```
I think you are missing a check for when the user hasn't set `LanguageClient_selectionUI`. As it stands, if that option is not set, but FZF is loaded, it will use the location list anyways. I would prefer to be explicit and have the user set it to FZF even if it was loaded, but I think we might be better of maintaining the current behaviour so we don't catch anyone by surprise with an unexpected change.

martskins commented on autoload/LanguageClient.vim:
```diff
@@ -1401,12 +1417,8 @@ endfunction
-                    \ 'source': l:options,
-                    \ 'sink': function('LanguageClient_handleContextMenuItem'),
-                    \ }))
+    if type(s:GetVar('LanguageClient_selectionUI')) is s:TYPE.funcref || s:GetVar('LanguageClient_selectionUI') ==? 'FZF'
```
Same here, if the option is not set at all but FZF is loaded it will use the location list, but the current behaviour is different.

martskins commented:
Nice work! Left a few comments, but it looks pretty good to me. I would probably add the config you shared here to use clap as an example or documentation, so it's easier for users to find that.

runiq committed:
Check for empty g:LanguageClient_selectionUI

...but still use FZF if it's installed.

runiq commented:
I've added the checks for `g:loaded_fzf`.

> I would probably add the config you shared here to use clap as an example or documentation, so it's easier for users to find that.

Right now this depends on @liuchengxu's [patch](https://github.com/liuchengxu/vim-clap/issues/433#issuecomment-646029971) which isn't on master yet; should I just keep this PR open until it's available?

martskins commented on autoload/LanguageClient.vim:
```diff
@@ -201,6 +209,15 @@ function! s:inputlist(...) abort
 
+function! s:selectionUI_funcref(source, sink) abort
+    if get(g:, 'LanguageClient_selectionUI', v:null) ==? 'FZF'
+                \ || get(g:, 'loaded_fzf')
```
I don't think this is what you want here. As it stands, if `loaded_fzf` is set then fzf will be used, regardless of what you have configured in `selectionUI`.

I think I would do something like `if get(g:, 'LanguageClient_selectionUI', 'FZF')  ==? 'FZF' && get(g: 'loaded_fzf')` instead, that way FZF will be the default but it will only be used if it's loaded. And if `selectionUI` is another thing, different than FZF, it will go to the else branch.

martskins commented on autoload/LanguageClient.vim:
```diff
@@ -1401,12 +1418,10 @@ endfunction
-                    \ }))
+    if type(get(g:, 'LanguageClient_selectionUI', v:null)) is s:TYPE.funcref
+                \ || get(g:, 'LanguageClient_selectionUI', v:null) ==? 'FZF'
+                \ || get(g:, 'loaded_fzf')
```
Same here, this will always fall in the if branch if fzf is loaded. I would do something like this.

```vim
if type(get(g:, 'LanguageClient_selectionUI', v:null)) is s:TYPE.funcref
    \ || (get(g:, 'LanguageClient_selectionUI', v:null) ==? 'FZF' && get(g:, 'loaded_fzf'))
```

runiq replied:
I just realized I completely dropped the `LanguageClient_fzfContextMenu` check. :flushed: So the final check would be:

```vim
if get(g:, 'LanguageClient_fzfContextMenu', 1)
    \ && (type(get(g:, 'LanguageClient_selectionUI', v:null)) is s:TYPE.funcref
        \ || (get(g:, 'LanguageClient_selectionUI', v:null) ==? 'FZF' && get(g:, 'loaded_fzf')))
```

This is just your version with the `LanguageClient_fzfContextMenu` check in front.
 I'm gonna add that back in, too, and alias it to `LanguageClient_selectionUIContextMenu` or something in a separate commit (naming suggestions welcome, of course).

martskins commented:
Added a few more comments on your last changes. As for the patch thing, even though this isn't exactly tied to clap, I would wait for them to merge that patch yeah.

Speaking of tied to clap, I'm not aware of any other selection UI of this sort, but it would be nice to see what else is there, to make sure that we have a generic enough design for this as to allow other plugins to make use of this.

runiq commented:
> Speaking of tied to clap, I'm not aware of any other selection UI of this sort

There's [skim](https://github.com/lotabout/skim) and its corresponding [Vim plugin](https://github.com/lotabout/skim.vim), at least, but that's pretty much just Clap without the RPC interface. Off the top of my head, I can only think of fzy, but I'm not at a computer at the moment, gotta tuck the young'un in. :)

runiq commented:
Alright, added in the changes. Sorry for messing that up. :(

runiq committed:
fzfContextMenu -> selectionUIContextMenu

runiq committed:
Update docs

runiq commented:
> Speaking of tied to clap, I'm not aware of any other selection UI of this sort

Here are the ones I've found:

Selector | `source` | `sink`
-|-|-
[fzf](https://github.com/junegunn/fzf) | List of strings \| Vim command as string | Funcref \| Vim command as string
[skim](https://github.com/lotabout/skim) | List of strings \| Vim command as string | Funcref \| Vim command as string
[vim-fzy](https://github.com/bfrg/vim-fzy) | List of strings \| Vim command as string | Funcref
[clap](https://github.com/liuchengxu/vim-clap) | List of strings \| Funcref \| Vim command as string | Funcref \| Vim command as string
[vim-picker](https://vimawesome.com/plugin/vim-picker-may-fear-less) | Shell command as string | Vim command as string
[fzy](https://github.com/jhawthorn/fzy) | None (just a shell command) | None (just a shell command)
[Denite](https://github.com/Shougo/denite.nvim) |  Python | Python
[LeaderF](https://github.com/Yggdroot/LeaderF) | Python | Python
[ctrlp](https://github.com/kien/ctrlp.vim/blob/extensions/autoload/ctrlp/sample.vim) | "The name of the input function including any arguments" as string | "The name of the output function" as string

I've found others, but they all appeared to be variations on the ones above WRT argument usage.

runiq commented:
Annnd again I realized that I should have based this PR on the dev branch. :( Will rebase.

Edit: Done.

runiq commented:
The [PR that implements `clap#run`](https://github.com/liuchengxu/vim-clap/pull/46) has been merged now.

martskins commented on autoload/LanguageClient.vim:
```diff
@@ -201,6 +209,15 @@ function! s:inputlist(...) abort
+    if get(g:, 'LanguageClient_selectionUI', 'FZF') ==? 'FZF'
+                \ && get(g:, 'loaded_fzf')
+        call s:FZF(a:source, a:sink)
+    else
```
There's still an issue here, sorry I didn't spot it before. If you set `let g:LanguageClient_selectionUI = 'location-list'` it will error, because it will jump to the else branch and that expects a `Funcref`.

martskins replied:
You'll probably need something similar to what's done in `contextMenu` which opens `inputlist` if it can't use FZF.

runiq replied:
Wow, this is turning out to be much more confusing than I had anticipated. Sorry *again* for messing up.

I guess I'll have to make a table of the possible interactions.

runiq replied:
Wait—that case shouldn't happen at all. `s:selectionUI_funcref()` is only called from the `SelectionUI::Funcref` codepaths in Rust, which would not be hit if `g:LanguageClient_selectionUI = 'location-list'` or `'quickfix'`. Right?

Edit: An error path seems warranted, though.

martskins replied:
Ah wait, it's an issue with code actions and code lens actions only. This is rather unfortunate, but maybe I can try fixing that before merging this one, let me see if I can push that through soon enough for you to rebase.

martskins commented:
On an **un**related note, the current `selectionUI` is not really respected everywhere, your PR fixes most of those places, as it calls that new function you implemented instead of directly calling `s:FZF` from Rust. But the contextMenu for example does still not respect that (I know this PR has nothing to do with this :smile: ), since it doesn't really check for the user config but rather it checks that fzf is loaded and uses it if it's available. No need to fix this in this PR, but figured I'd put it in here in case you end up touching that section of the code and come up with a fix for that.

runiq committed:
Make s:selectionUI_funcref more robust

Now shows a meaningful error when g:LanguageClient_selectionUI is not
'FZF' or a funcref.

runiq commented:
> On an **un**related note, the current `selectionUI` is not really respected everywhere, your PR fixes most of those places, as it calls that new function you implemented instead of directly calling `s:FZF` from Rust. But the contextMenu for example does still not respect that (I know this PR has nothing to do with this smile ), since it doesn't really check for the user config but rather it checks that fzf is loaded and uses it if it's available. No need to fix this in this PR, but figured I'd put it in here in case you end up touching that section of the code and come up with a fix for that.

There's actually `g:LanguageClient_fzfContextMenu` which decides whether FZF or the input list is used for the context menu (FZF is the default, though).

I feel like the selection UI handling is already pretty convoluted at the moment: It needs to deal with backwards compatibility concerns, the Rust—Vim interface, and graceful fallback to sane values in the face of adversity. I'd rather punt this problem to a separate PR if that's okay with you :)

martskins commented:
> > On an **un**related note, the current `selectionUI` is not really respected everywhere, your PR fixes most of those places, as it calls that new function you implemented instead of directly calling `s:FZF` from Rust. But the contextMenu for example does still not respect that (I know this PR has nothing to do with this smile ), since it doesn't really check for the user config but rather it checks that fzf is loaded and uses it if it's available. No need to fix this in this PR, but figured I'd put it in here in case you end up touching that section of the code and come up with a fix for that.
> 
> There's actually `g:LanguageClient_fzfContextMenu` which decides whether FZF or the input list is used for the context menu (FZF is the default, though).
> 
> I feel like the selection UI handling is already pretty convoluted at the moment: It needs to deal with backwards compatibility concerns, the Rust—Vim interface, and graceful fallback to sane values in the face of adversity. I'd rather punt this problem to a separate PR if that's okay with you :)

Sure thing! :+1:

martskins commented:
@runiq opened #1060 to abstract the selection UI presenting bit. The rebase looks a little painful, but happy to lend a hand if you need it. That fixes the selection UI used by codeLenses and codeActions, so that users that choose to migrate from fzf to clap after this PR don't loose that feature.

I opened it as a draft, as I don't want that to get merged without @autozimu having a look at it. The new function names I'm not 100% happy about, so if you can think of something better, happy to takke suggestions. :smile:

runiq commented:
Rebased!

So there appear to be two remaining issues: Your [comment above](https://github.com/autozimu/LanguageClient-neovim/pull/1059#discussion_r443008140) about code actions/lenses which should hopefully be addressed now, and the context menu thing:

>> […] the current `selectionUI` is not really respected everywhere […] the contextMenu for example […] doesn't really check for the user config but rather it checks that fzf is loaded and uses it if it's available.
>
> There's actually `g:LanguageClient_fzfContextMenu` which decides whether FZF or the input list is used for the context menu (FZF is the default, though). […] I'd rather punt this problem to a separate PR if that's okay with you :)

After thinking about this, I wouldn't mind including it in this PR after all. :man_shrugging: Using `present_actions()` from the context menu would involve somehow piping through `LanguageClient_contextMenu()` from Vim through to the Rust side, which means adding a new `LanguageClient/presentActions` request if I understand correctly? I'll have to look into this a bit.

I'm a fan of it, though! Always nice to streamline stuff like this. :)

Edit: Actually, it'd probably be a good idea to create those `LanguageClient/present{Actions,List}` requests regardless of where they are needed. Right now, the `present_{list,actions}()` functions are single points of responsibility from the Rust side, but I'm handling option shuffling in multiple points on the Vim side. That feels suboptimal now that #1060 has been merged. I'd rather like for the Vim side to be able to somehow hook into `present_{list,actions}()` as well.

Thoughts?

runiq commented:
Hmm, I must have messed something up while rebasing. Code actions and lenses don't appear to work—I can select them in Clap, but the action is not triggered. Will investigate.

martskins commented:
Actually, I completely overlooked something about the context menu, which I found while opening that other PR. There's `g:LanguageClient_fzfContextMenu`, which defaults to `1`, and overrides your `g:LanguageClient_selectionUI` selection, so that explains why `contextMenu` ignores that :smile: . 
Considering that, I'd probably leave it as is, as changing that (even if it was to a "more correct") behaviour would "break" stuff for some users. We can address that eventually though, but it would basically mean removing that `g:LanguageClient_fzfContextMenu` config.

**EDIT**: I just noticed you added a `selectionUIContextMenu` config, that's fine, but we should probably still keep the `fzfContextMenu` around.

runiq commented:
> I just noticed you added a `selectionUIContextMenu` config, that's fine, but we should probably still keep the `fzfContextMenu` around.

I do keep it around—the code recognizes both, and the help tag for both points to the same doc entry. They just mention that `fzfContextMenu` is deprecated.

martskins commented:
> Hmm, I must have messed something up while rebasing. Code actions and lenses don't appear to work—I can select them in Clap, but the action is not triggered. Will investigate.

Fwiw, it seems to be working for me.

martskins commented:
> > I just noticed you added a `selectionUIContextMenu` config, that's fine, but we should probably still keep the `fzfContextMenu` around.
> 
> I do keep it around—the code recognizes both, and the help tag for both points to the same doc entry. They just mention that `fzfContextMenu` is deprecated.

Ah yeah sorry, I missed that.

runiq commented:
> Fwiw, it seems to be working for me.

Weird. I rebased on top of the base branch, but I'm still not able to trigger *either* code actions or code lenses in Clap. Spent half a day yesterday trying to make things to work, but I had to go deeper and deeper into the rabbit hole—I stopped when I appeared to hit an error in `Runnable::deserialize()`. :( This is the patch that allows me to run code actions at least, but not code lenses:

```diff
diff --git i/src/language_server_protocol.rs w/src/language_server_protocol.rs
index 77b104b..262cbec 100644
--- i/src/language_server_protocol.rs
+++ w/src/language_server_protocol.rs
@@ -1992,9 +1992,9 @@ impl LanguageClient {
             .map(|cl| match &cl.command {
                 None => Err(anyhow!("no command, skipping")),
                 Some(cmd) => Ok(CodeAction {
-                    kind: Some(cmd.title.clone()),
-                    title: cmd.command.clone(),
-                    command: cl.clone().command,
+                    kind: Some(cmd.command.clone()),
+                    title: cmd.title.clone(),
+                    command: cl.command.clone(),
                     diagnostics: None,
                     edit: None,
                     is_preferred: None,
@@ -3418,14 +3418,14 @@ impl LanguageClient {
         let selection: String =
             try_get("selection", params)?.ok_or_else(|| anyhow!("selection not found!"))?;
         let tokens: Vec<&str> = selection.splitn(2, ": ").collect();
-        let title = tokens
+        let kind = tokens
             .get(0)
             .cloned()
-            .ok_or_else(|| anyhow!("Failed to get title! tokens: {:?}", tokens))?;
-        let kind = tokens
+            .ok_or_else(|| anyhow!("Failed to get kind! tokens: {:?}", tokens))?;
+        let title = tokens
             .get(1)
             .cloned()
-            .ok_or_else(|| anyhow!("Failed to get kind! tokens: {:?}", tokens))?;
+            .ok_or_else(|| anyhow!("Failed to get title! tokens: {:?}", tokens))?;
         let actions = self.get(|state| state.stashed_code_action_actions.clone())?;
         let idx = actions
             .iter()
diff --git i/src/types.rs w/src/types.rs
index e82c532..aaa6baa 100644
--- i/src/types.rs
+++ w/src/types.rs
@@ -1252,6 +1252,7 @@ pub struct VirtualText {
 }

 #[derive(Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
 pub struct WorkspaceEditWithCursor {
     pub workspace_edit: WorkspaceEdit,
     pub cursor_position: Option<TextDocumentPositionParams>,
```

Not sure what's going on here, to be honest.

Are you using FZF as your selection UI?

martskins commented:
Ah, good catch on that missing `rename_all`. I'll push a fix for that in a few minutes.

About the selection UI, I tried FZF and location-list in dev. And I tried both of those and clap in this branch. What are the code lenses you are trying to run? Which language and what do they do?

runiq commented:
I'm using Rust-analyzer and [this tiny program for testing](https://git.sr.ht/~runiq/seddate/tree/master/src/main.rs). I'm trying to trigger the "Run" codelens from [`fn main()`](https://git.sr.ht/~runiq/seddate/tree/master/src/main.rs#L50).

[These are the interesting bits from a debug log file](https://gist.github.com/runiq/4e7ebdcba6406ded170db0805be564ab#file-lcn-seddate-log-L10-L19). I suspect I'm hitting some edge case from [this recent PR](https://github.com/autozimu/LanguageClient-neovim/pull/1058).

martskins commented:
Which version of rust-analyzer are you running? `rust-analyzer version` yields `rust-analyzer f4f5117` for me.

runiq commented:
I had an old version, but the error is the same with the [most recent weekly release](https://github.com/rust-analyzer/rust-analyzer/releases/tag/2020-06-22). The log file was created with that newest release of rust-analyzer.

Edit: The result from `rust-analyzer --version` is `rust-analyzer 31611da`.

martskins commented:
Ah yeah, I can reproduce now. Seems like there were some other changes as well. This is unfortunate. Let me try and fix that and I'll come back to this. This is not related to this PR though, but it's worth fixing asap anyways.

runiq commented:
No sweat, shaving yaks is what we do :)

runiq commented:
I've updated the [log file Gist](https://gist.github.com/runiq/4e7ebdcba6406ded170db0805be564ab)  with a reformatted version of the stashed code actions in line 10; I thought it might be useful for debugging serialization errors.

martskins commented:
@runiq I've opened #1071. That should allow running both the old and new version of rust-analyzer runnables. It also fixes the issue you found, it seems like I was trying a very narrow set of tests and those were passing but I eventually found the issue you had. I'll merge it in a few minutes anyways, as it's currently broken, so it can only get better :smile:

runiq commented:
Great, thanks, rebased! Looks like code lenses and actions are triggered just fine now. (I've found some separate small issues, but those are rust-analyzer-related, I believe. I'll file separate issues for those)

<details><summary>Remaining issues (not related to this PR)</summary>
Edit: Spoke too soon. Two remaining issues:

- Applying a code action doesn't reposition the cursor, even though `applyWorkspaceEdit()` appears to finish without errors.
- There seems to be an issue with correct window focus—it's even apparent in the [video in the OP](https://user-images.githubusercontent.com/620981/85034923-e4295400-b182-11ea-9ac1-5a9852ff0a44.gif): After selecting a location, the window doesn't update the line number display on the left, like it's supposed to.

I'll have a look, though I'm not entirely sure yet whether that's an issue with my config, or this PR, or something else…

Edit2: The first issue doesn't seem to have anything to do with this PR, since `applyWorkspaceEdit()` always reports the same position *after* the edit as *before*. Which, [according to the RA manual](https://rust-analyzer.github.io/manual.html#add_impl), shouldn't happen for the actions I tried (impl Struct, add custom impl). I'll file a bug.

Edit3: The second issue is with my vim-numbertoggle installation, and doesn't have anything to do with this either.
</details>

So, at least from my point of view, this PR is finally good to go.

martskins commented:
I don't fully understand the first issue you mentioned, and I haven't noticed anything weird, but please do file an issue about this so we can track it, if you can add a recording of what's wrong with it it would be great.

As for the PR itself, I'll give this a ride today and see how it behaves, but I'm pretty sure it should be fine now. So if everything works fine I'll merge this at the end(ish) of the day.

Thank you for implementing this. Sorry it took this long and you thank you for bearing with me :smile:

runiq commented:
> I don't fully understand the first issue you mentioned

If I have this struct (the `|` is the cursor position):

```rust
struct Foo {
bar: u32,|
}
```

and I run the "Implement Foo" code action, I end up with this, which is correct code, but the cursor is in the wrong position (it's supposed to be in the new `impl` block, as per the [RA manual](https://rust-analyzer.github.io/manual.html#add_impl)):

```rust
struct Foo {
bar: u32,|
}

impl Foo {
}
```

martskins commented:
Ah ok, yeah, that's an experimental feature by rust-analyzer, which is not supported by all clients. See [here](https://github.com/rust-analyzer/rust-analyzer/blob/1bc1f28bc58b2dbcf8f8f548c277e2c90e3075cd/crates/rust-analyzer/src/config.rs#L298).

runiq commented:
Ah, right, sorry—can't say I fully understand the architecture yet, TBH. :sweat:
